docs/config.md
# Documentation on config module
This document will describe the config.py module

## Overview
- This module reads all user settings and configures the python modules to process according to these user inputs.
- The user inputs are read from the env file that is generated with [enveditor](https://github.com/ncsa/ameriflux-pipeline/blob/develop/docs/enveditor.md) module.
- If some required configurations are not provided by the user, this module sets a default value.
- This module also adds additional configurations that are not user-defined.
  - QC_PRECIP_LOWER gives the precipitation lower threshold value in inches. This is set as 0.0
  - QC_PRECIP_UPPER gives the precipitation upper threshold value in inches. This is set as 0.2
    - User can modify these settings [here](https://github.com/ncsa/ameriflux-pipeline/blob/develop/ameriflux_pipeline/config.py#L132).
  - MET_TIMEPERIOD gives the timeperiod (in minutes) of one record in meteorological data file. This is set as 30.0
  - PRECIP_TIMEPERIOD gives the timeperiod (in minutes) of one record in percipitation data file. This is set as 5.0.
    - User can modify these settings [here](https://github.com/ncsa/ameriflux-pipeline/blob/develop/ameriflux_pipeline/config.py#L137).
  - PYFLUXPRO_OVERLAP_TIMESTAMP flag checks if eddypro fulloutput sheet and metdata sheet in pyfluxpro_input.xlsx sheet need to have overlapping timestamps.
    - If set to true, eddypro fulloutput sheet and metdata sheet needs to have overlapping timestamps. If not, the pyfluxpro data processing will be aborted.
    - If set to False, the check for overlapping timestamp will not be executed.
    - User can modify these settings [here](https://github.com/ncsa/ameriflux-pipeline/blob/develop/ameriflux_pipeline/config.py#L142).
- Users can change the configuration settings by modifying the [config](https://github.com/ncsa/ameriflux-pipeline/blob/develop/ameriflux_pipeline/config.py) module.
- The default values can be changed by modifying the second parameter in ```os.getenv()``` function for the corresponding settings.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - 11-30-2022

### Added
- Data preprocessing completed for creation of master met data file as per guide. [#1](https://github.com/ncsa/ameriflux-pipeline/issues/1)
- Documentation for automating meteorological data. [#8](https://github.com/ncsa/ameriflux-pipeline/issues/8)
- Created config to parse env variables. [#11](https://github.com/ncsa/ameriflux-pipeline/pull/11)
- Data formatting completed for EddyPro as per guide. [#12](https://github.com/ncsa/ameriflux-pipeline/issues/12)
- PyTest for preprocessor. [#13](https://github.com/ncsa/ameriflux-pipeline/issues/13)
- Automation for running EddyPro. [#21](https://github.com/ncsa/ameriflux-pipeline/issues/21)
- Data formatting for PyFluxPro as per guide. [#25](https://github.com/ncsa/ameriflux-pipeline/issues/25)
- CSV comparison script. [#27](https://github.com/ncsa/ameriflux-pipeline/issues/27)
- QA/QC check for precipitation data. [#39](https://github.com/ncsa/ameriflux-pipeline/issues/39)
- Created NOTES.md to document code structure and decisions made. [#47](https://github.com/ncsa/ameriflux-pipeline/issues/47)
- Automation chain for running eddypro. [#50](https://github.com/ncsa/ameriflux-pipeline/issues/50)
- Env variable for automating user confirmation to insert large number of missing timestamps. [#52](https://github.com/ncsa/ameriflux-pipeline/issues/52)
- Merging of met data and precip data with different timestamps. [#54](https://github.com/ncsa/ameriflux-pipeline/issues/54)
- Automated creation of pyfluxpro input excel spreadsheet [#55](https://github.com/ncsa/ameriflux-pipeline/issues/55)
- EddyPro automation separation for Mac and Windows. [#56](https://github.com/ncsa/ameriflux-pipeline/issues/56)
- Data sync between remote sftp server and local machine. [#57](https://github.com/ncsa/ameriflux-pipeline/issues/57)
- Restructured code to have data directory and separate directories for master met, eddypro and pyfluxpro. [#68](https://github.com/ncsa/ameriflux-pipeline/issues/68)
- GUI application for editing and creating .env file [#77](https://github.com/ncsa/ameriflux-pipeline/issues/77)
- Reconstructed code to handle Maize site data (ZMB) [#78](https://github.com/ncsa/ameriflux-pipeline/issues/78)
- Pyfluxpro input sheet formatting for AmeriFlux submission [#88](https://github.com/ncsa/ameriflux-pipeline/issues/88)
- Creation of L1.txt meeting AmeriFlux standards [#92](https://github.com/ncsa/ameriflux-pipeline/issues/92)
- Added missing variables to L1.txt for Ameriflux [#99](https://github.com/ncsa/ameriflux-pipeline/issues/99)
- L1 processing is made more efficient [#100](https://github.com/ncsa/ameriflux-pipeline/issues/100)
- AmeriFlux L1 formatting variables to env editor [#108](https://github.com/ncsa/ameriflux-pipeline/issues/108)
- Added user confirmation to replace certain PyFluxPro variable names to AmeriFlux names [#105](https://github.com/ncsa/ameriflux-pipeline/issues/105)
- Check formatting of L1 inputs [#109](https://github.com/ncsa/ameriflux-pipeline/issues/109)
- Updating units for Ameriflux only variables [#115](https://github.com/ncsa/ameriflux-pipeline/issues/115)
- AmeriFlux L1 erroring variables to env editor [#117](https://github.com/ncsa/ameriflux-pipeline/issues/117)
- Creation of L2.txt meeting Ameriflux standards [#101](https://github.com/ncsa/ameriflux-pipeline/issues/101)
- L2 AmeriFlux formatting variables to env editor. [#126](https://github.com/ncsa/ameriflux-pipeline/issues/126)
- Formatting of PyFluxPro output for Ameriflux. [#130](https://github.com/ncsa/ameriflux-pipeline/issues/130)
- Reading precip data more robust. [#137](https://github.com/ncsa/ameriflux-pipeline/issues/137)
- Merging to various raw met data files to csv format for a specified time period. [#134](https://github.com/ncsa/ameriflux-pipeline/issues/134)
- Tested and updated pipeline for MaizeBasalt site [#154](https://github.com/ncsa/ameriflux-pipeline/issues/154)
- Correction of fill values and filename of final Ameriflux csv file [#162](https://github.com/ncsa/ameriflux-pipeline/issues/162)
- User input validation [#159](https://github.com/ncsa/ameriflux-pipeline/issues/159)
- Updated NOTES [#166](https://github.com/ncsa/ameriflux-pipeline/issues/166)
- Reading xls formatting functionality [#164](https://github.com/ncsa/ameriflux-pipeline/issues/164)
- Details on met_data_merge in README [#178](https://github.com/ncsa/ameriflux-pipeline/issues/178)
- Data validation for met data merge [#168](https://github.com/ncsa/ameriflux-pipeline/issues/168)
- Data validation for creation of master met data [#172](https://github.com/ncsa/ameriflux-pipeline/issues/172)
- Data validation for formatting master met data for EddyPro [#173](https://github.com/ncsa/ameriflux-pipeline/issues/173)
- Data validation for creating pyfluxpro input excel sheet [#183](https://github.com/ncsa/ameriflux-pipeline/issues/183)
- Data validation for formatting pyfluxpro input excel sheet for ameriflux [#186](https://github.com/ncsa/ameriflux-pipeline/issues/186)
- Data validation for formatting L1 control file for pyfluxpro. [#187](https://github.com/ncsa/ameriflux-pipeline/issues/187)
- Data validation for formatting L2 control file for pyfluxpro. [#188](https://github.com/ncsa/ameriflux-pipeline/issues/188)
- Data validation for formatting pyfluxpro output for Ameriflux submission. [#203](https://github.com/ncsa/ameriflux-pipeline/issues/203)
- Created logging for met_data_merge. [#15](https://github.com/ncsa/ameriflux-pipeline/issues/15)
- Logging for pre_pyfluxpro. [#207](https://github.com/ncsa/ameriflux-pipeline/issues/207)
- Logging for utils. [#209](https://github.com/ncsa/ameriflux-pipeline/issues/209)
- Logging for mastermet. [#211](https://github.com/ncsa/ameriflux-pipeline/issues/211)

ameriflux_pipeline/config.py
# Copyright (c) 2022 University of Illinois and others. All rights reserved.
#
# This program and the accompanying materials are made available under the
# terms of the Mozilla Public License v2.0 which accompanies this distribution,
# and is available at https://www.mozilla.org/en-US/MPL/2.0/

# configs file
import os
from dotenv import load_dotenv

# Load .env file
load_dotenv()


class Config:
    """
    class to list all configuration settings required for preprocessing and formatting for EddyPro and PyFluxPro
    """
    # obtaining ghg files using rsync
    # user confirmation to perform rsync
    SFTP_CONFIRMATION = os.getenv('SFTP_CONFIRMATION', 'N')
    SFTP_SERVER = os.getenv('SFTP_SERVER')
    SFTP_USERNAME = os.getenv('SFTP_USERNAME')
    SFTP_PASSWORD = os.getenv('SFTP_PASSWORD')
    SFTP_GHG_REMOTE_PATH = os.getenv('SFTP_GHG_REMOTE_PATH')
    SFTP_GHG_LOCAL_PATH = os.getenv('SFTP_GHG_LOCAL_PATH')
    SFTP_MET_REMOTE_PATH = os.getenv('SFTP_MET_REMOTE_PATH')
    SFTP_MET_LOCAL_PATH = os.getenv('SFTP_MET_LOCAL_PATH')

    # input data for creating master meteorology data
    # input met data path
    INPUT_MET = os.getenv('INPUT_MET',
                          '/Users/ameriflux-pipeline/ameriflux_pipeline/data/'
                          'master_met/input/FLUXSB_EC_JanMar2021.csv')
    # input precipitation data
    INPUT_PRECIP = os.getenv('INPUT_PRECIP',
                             '/Users/ameriflux-pipeline/ameriflux_pipeline/data/'
                             'master_met/input/Precip_IWS_Jan-Feb_2021.xlsx')
    # Number of missing timeslot threshold for user confirmation to insert
    MISSING_TIME = os.getenv('MISSING_TIME', 96)
    # NOTE 9
    # User confirmation to insert large number of missing timestamps
    # Enter 'Y'/'YES' to insert, 'N'/'NO' to ignore and 'A'/'ASK' to await user input during runtime.
    MISSING_TIME_USER_CONFIRMATION = os.getenv('MISSING_TIME_USER_CONFIRMATION', 'Y')
    # master met output data path
    MASTER_MET = os.getenv('MASTER_MET',
                           '/Users/ameriflux-pipeline/ameriflux_pipeline/data/'
                           'master_met/output/met_output.csv')

    # input data for formatting Eddypro master meteorology data.
    # input soil key data path
    INPUT_SOIL_KEY = os.getenv('INPUT_SOIL_KEY',
                               '/Users/ameriflux-pipeline/ameriflux_pipeline/data/'
                               'eddypro/input/Soils_key.xlsx')

    # eddypro related parameters
    # bin folder location of eddypro_rp exec file
    EDDYPRO_BIN_LOC = os.getenv('EDDYPRO_BIN_LOC', '')  # '/Applications/eddypro.app/Contents/MacOS/bin'
    # EDDYPRO_PROJ_FILE_TEMPLATE
    EDDYPRO_PROJ_FILE_TEMPLATE = os.getenv('EDDYPRO_PROJ_FILE_TEMPLATE', '/Users/ameriflux-pipeline/ameriflux_pipeline/'
                                                                         'eddypro/templates/template.eddypro')
    # EddyPro run template
    EDDYPRO_PROJ_FILE_NAME = os.getenv('EDDYPRO_PROJ_FILE_NAME',
                                       '/Users/ameriflux-pipeline/ameriflux_pipeline/data/'
                                       'eddypro/templates/ameriflux.eddypro')
    EDDYPRO_PROJ_TITLE = os.getenv('EDDYPRO_PROJ_TITLE', 'AmeriFlux_Pipeline')
    EDDYPRO_PROJ_ID = os.getenv('EDDYPRO_PROJ_ID', 'ameriflux_pipeline')
    EDDYPRO_FILE_PROTOTYPE = os.getenv('EDDYPRO_FILE_PROTOTYPE', 'yyyy-mm-ddTHHMM??_Sorghum-00137.ghg')
    # EddyPro input metadata file from GHG
    EDDYPRO_PROJ_FILE = os.getenv('EDDYPRO_PROJ_FILE',
                                  '/Users/ameriflux-pipeline/ameriflux_pipeline/data/'
                                  'eddypro/input/2021-01-01T000000_Sorghum-00137.metadata')
    # EddyPro input dynamic metadata file
    EDDYPRO_DYN_METADATA = os.getenv('EDDYPRO_DYN_METADATA',
                                     '/Users/ameriflux-pipeline/ameriflux_pipeline/data/'
                                     'eddypro/input/Sorghum_2021_dynamic_metadata.csv')
    # EddyPro input GHG files folder
    EDDYPRO_INPUT_GHG_PATH = os.getenv('EDDYPRO_INPUT_GHG_PATH',
                                       '/Users/ameriflux-pipeline/ameriflux_pipeline/data/'
                                       'eddypro/input/ghg/')
    # EddyPro output folder
    EDDYPRO_OUTPUT_PATH = os.getenv('EDDYPRO_OUTPUT_PATH',
                                    '/Users/ameriflux-pipeline/ameriflux_pipeline/data/eddypro/output/')

    # PyFluxPro related data
    FULL_OUTPUT_PYFLUXPRO = os.getenv('FULL_OUTPUT_PYFLUXPRO',
                                      '/Users/ameriflux-pipeline/ameriflux_pipeline/data/pyfluxpro/'
                                      'generated/full_output.csv')
    MET_DATA_30_PYFLUXPRO = os.getenv('MET_DATA_30_PYFLUXPRO',
                                      '/Users/ameriflux-pipeline/ameriflux_pipeline/data/pyfluxpro/'
                                      'generated/Met_data_30.csv')
    PYFLUXPRO_INPUT_SHEET = os.getenv('PYFLUXPRO_INPUT_SHEET',
                                      '/Users/ameriflux-pipeline/ameriflux_pipeline/data/pyfluxpro/'
                                      'generated/pyfluxpro_input.xlsx')
    PYFLUXPRO_INPUT_AMERIFLUX = os.getenv('PYFLUXPRO_INPUT_AMERIFLUX',
                                          '/Users/ameriflux-pipeline/ameriflux_pipeline/data/pyfluxpro/'
                                          'generated/pyfluxpro_input_ameriflux.xlsx')
    # L1 Ameriflux Formatting
    L1_MAINSTEM_INPUT = os.getenv('L1_MAINSTEM_INPUT',
                                  '/Users/ameriflux-pipeline/ameriflux_pipeline/data/pyfluxpro/input/'
                                  'L1_mainstem.txt')
    L1_AMERIFLUX_ONLY_INPUT = os.getenv('L1_AMERIFLUX_ONLY_INPUT', '/Users/ameriflux-pipeline/ameriflux_pipeline/'
                                                                   'data/pyfluxpro/input/L1_ameriflux_only.txt')
    L1_AMERIFLUX_MAINSTEM_KEY = os.getenv('L1_AMERIFLUX_MAINSTEM_KEY',
                                          '/Users/ameriflux-pipeline/ameriflux_pipeline/data/pyfluxpro/'
                                          'input/Ameriflux-Mainstem-Key.xlsx')
    L1_AMERIFLUX_RUN_OUTPUT = os.getenv('L1_AMERIFLUX_RUN_OUTPUT',
                                        '/Users/ameriflux-pipeline/ameriflux_pipeline/data/pyfluxpro/'
                                        'generated/Sorghum_2021_L1.nc')
    L1_AMERIFLUX = os.getenv('L1_AMERIFLUX',
                             '/Users/ameriflux-pipeline/ameriflux_pipeline/data/pyfluxpro/generated/L1_ameriflux.txt')

    # User confirmation to replace erroring Ameriflux variable names in L1
    # Enter 'Y'/'YES' to replace, 'N'/'NO' to ignore and 'A'/'ASK' to await user input during runtime.
    AMERIFLUX_VARIABLE_USER_CONFIRMATION = os.getenv('AMERIFLUX_VARIABLE_USER_CONFIRMATION', 'N')
    L1_AMERIFLUX_ERRORING_VARIABLES_KEY = os.getenv('L1_AMERIFLUX_ERRORING_VARIABLES_KEY',
                                                    '/Users/ameriflux-pipeline/ameriflux_pipeline/data/pyfluxpro/'
                                                    'input/L1_erroring_variables.xlsx')

    # L2 Ameriflux Formatting
    L2_MAINSTEM_INPUT = os.getenv('L2_MAINSTEM_INPUT', '/Users/ameriflux-pipeline/ameriflux_pipeline/data/'
                                                       'pyfluxpro/input/L2_mainstem.txt')
    L2_AMERIFLUX_ONLY_INPUT = os.getenv('L2_AMERIFLUX_ONLY_INPUT', '/Users/ameriflux-pipeline/ameriflux_pipeline/data/'
                                                                   'pyfluxpro/input/L2_ameriflux_only.txt')
    L2_AMERIFLUX_RUN_OUTPUT = os.getenv('L2_AMERIFLUX_RUN_OUTPUT',
                                        '/Users/ameriflux-pipeline/ameriflux_pipeline/data/'
                                        'pyfluxpro/generated/Sorghum_2021_L2.nc')
    L2_AMERIFLUX = os.getenv('L2_AMERIFLUX',
                             '/Users/ameriflux-pipeline/ameriflux_pipeline/data/pyfluxpro/generated/L2_ameriflux.txt')

    # QA/QC values
    # precipitation threshold values used in creating master meteorological data
    QC_PRECIP_LOWER = 0.0  # precipitation lower threshold value (inches)
    QC_PRECIP_UPPER = 0.2  # precipitation upper threshold value (inches)

    # Data timeperiods
    # set meteorological data timeperiod
    MET_TIMEPERIOD = 30.0
    # set precipitation data timeperiod
    PRECIP_TIMEPERIOD = 5.0

    # PyFluxpro overlap timestamp
    # flag to check if fulloutput and metdata sheet in pyfluxpro_input.xlsx sheet has overlapping timestamps
    PYFLUXPRO_OVERLAP_TIMESTAMP = True  # setting to true checks for overlap
